File: tests/fixtures-support/loaders/fixture_envelope.py
"""Cross-language fixture envelope loader (Python reference implementation).

Validation rules are mirrored in ``fixture-envelope.ts`` and
``loaders/rust-loader/``. A rule change must land in all three at once.
Envelope contract: ``packages/schemas/fixture-envelope.v1.json`` (unified,
ADR-017; these loaders enforce the dataset-variant rules of that schema).
Payload validation is intentionally shallow (required + const at the top
level of the referenced canonical schema); deep validation belongs to the
schema toolchain, not to fixtures.
"""

from __future__ import annotations

import argparse
import json
import re
import sys
from pathlib import Path
from typing import Any, Dict, List, Tuple

ENVELOPE_VERSION = "1.0"
FIXTURE_ID_RE = re.compile(r"^[a-z0-9]+(-[a-z0-9]+)*(\/[a-z0-9]+(-[a-z0-9]+)*)*$")
NAME_RE = re.compile(r"^[a-z][a-z0-9-]*$")
TOP_LEVEL_KEYS = (
    "fixture_envelope_version",
    "fixture_id",
    "kind",
    "description",
    "input",
    "expected_outputs",
    "provenance",
    "stability",
)
REQUIRED_TOP_LEVEL_KEYS = ("fixture_envelope_version", "fixture_id", "kind", "input", "provenance")
INPUT_KEYS = ("schema", "payload")
OUTPUT_KEYS = ("name", "schema", "payload")
PROVENANCE_KEYS = ("origin", "synthetic", "license", "notes")
REQUIRED_PROVENANCE_KEYS = ("origin", "synthetic")
ORIGINS = ("synthetic", "curated-public", "user-contributed")
STABILITY_KEYS = ("deterministic", "ignored_fields")


class EnvelopeError(Exception):
    """Validation failure carrying a stable rule code for test attribution."""

    def __init__(self, rule: str, message: str) -> None:
        super().__init__(f"[{rule}] {message}")
        self.rule = rule


def _expect_object(value: Any, rule: str, what: str) -> Dict[str, Any]:
    if not isinstance(value, dict):
        raise EnvelopeError(rule, f"{what} must be a JSON object")
    return value


def base_schema_name(schema_id: str) -> str:
    """Strip the ``.json`` extension and trailing ``.vN`` from a schema $id segment."""
    segment = [part for part in schema_id.split("/") if part][-1]
    segment = re.sub(r"\.json$", "", segment)
    return re.sub(r"\.v\d+$", "", segment)


def load_schema_registry(schemas_dir: Path) -> Dict[str, Tuple[Path, Dict[str, Any]]]:
    registry: Dict[str, Tuple[Path, Dict[str, Any]]] = {}
    for path in sorted(Path(schemas_dir).glob("*.json")):
        parsed = json.loads(path.read_text(encoding="utf-8"))
        schema_id = parsed.get("$id")
        if isinstance(schema_id, str):
            registry[schema_id] = (path, parsed)
    return registry


def _validate_payload_shallow(payload: Any, schema: Dict[str, Any], schema_id: str, what: str) -> None:
    obj = _expect_object(payload, "payload-shape", what)
    for key in schema.get("required", []):
        if isinstance(key, str) and key not in obj:
            raise EnvelopeError("payload-required", f'{what} is missing required field "{key}" of {schema_id}')
    properties = schema.get("properties")
    if isinstance(properties, dict):
        for key, definition in properties.items():
            if key not in obj or not isinstance(definition, dict):
                continue
            constant = definition.get("const")
            if constant is not None and obj[key] != constant:
                raise EnvelopeError(
                    "payload-const",
                    f'{what} field "{key}" must equal {json.dumps(constant)} per {schema_id}',
                )


def _validate_resolvable_payload(container: Dict[str, Any], registry: Dict[str, Any], what: str) -> None:
    schema_id = container.get("schema")
    if not isinstance(schema_id, str) or not schema_id:
        raise EnvelopeError("input-shape", f"{what}.schema must be a non-empty string")
    entry = registry.get(schema_id)
    if entry is None:
        raise EnvelopeError("schema-unresolved", f'{what}.schema "{schema_id}" is not registered in packages/schemas')
    _validate_payload_shallow(container.get("payload"), entry[1], schema_id, f"{what}.payload")


def validate_envelope(raw: Any, fixture_dir_name: str, registry: Dict[str, Any]) -> Dict[str, Any]:
    envelope = _expect_object(raw, "shape", "fixture envelope")
    for key in envelope:
        if key not in TOP_LEVEL_KEYS:
            raise EnvelopeError("unknown-field", f'unknown top-level field "{key}"')
    for key in REQUIRED_TOP_LEVEL_KEYS:
        if key not in envelope:
            raise EnvelopeError("missing-field", f'missing required field "{key}"')
    if envelope["fixture_envelope_version"] != ENVELOPE_VERSION:
        raise EnvelopeError(
            "envelope-version",
            f'fixture_envelope_version must be "{ENVELOPE_VERSION}", got {json.dumps(envelope["fixture_envelope_version"])}',
        )
    fixture_id = envelope["fixture_id"]
    if not isinstance(fixture_id, str) or not FIXTURE_ID_RE.match(fixture_id):
        raise EnvelopeError("fixture-id", f'fixture_id {json.dumps(fixture_id)} is not kebab-case')
    if fixture_id != fixture_dir_name:
        raise EnvelopeError("fixture-id", f'fixture_id "{fixture_id}" must equal its directory name "{fixture_dir_name}"')
    kind = envelope["kind"]
    if not isinstance(kind, str) or not NAME_RE.match(kind):
        raise EnvelopeError("kind", f"kind {json.dumps(kind)} is not a lowercase kebab-case name")
    inp = _expect_object(envelope["input"], "input-shape", "input")
    for key in inp:
        if key not in INPUT_KEYS:
            raise EnvelopeError("input-shape", f'unknown input field "{key}"')
    for key in INPUT_KEYS:
        if key not in inp:
            raise EnvelopeError("input-shape", f'input is missing "{key}"')
    if isinstance(inp["schema"], str) and base_schema_name(inp["schema"]) != kind:
        raise EnvelopeError(
            "kind-mismatch",
            f'kind "{kind}" does not match input schema base name "{base_schema_name(inp["schema"])}"',
        )
    _validate_resolvable_payload(inp, registry, "input")
    outputs = envelope.get("expected_outputs")
    if outputs is not None:
        if not isinstance(outputs, list):
            raise EnvelopeError("output-shape", "expected_outputs must be an array")
        for output in outputs:
            entry = _expect_object(output, "output-shape", "expected_outputs entry")
            for key in entry:
                if key not in OUTPUT_KEYS:
                    raise EnvelopeError("output-shape", f'unknown expected_outputs field "{key}"')
            for key in OUTPUT_KEYS:
                if key not in entry:
                    raise EnvelopeError("output-shape", f'expected_outputs entry is missing "{key}"')
            if not isinstance(entry["name"], str) or not NAME_RE.match(entry["name"]):
                raise EnvelopeError("output-shape", f'expected_outputs name {json.dumps(entry["name"])} is invalid')
            _validate_resolvable_payload(entry, registry, f'expected_outputs "{entry["name"]}"')
    provenance = _expect_object(envelope["provenance"], "provenance-shape", "provenance")
    for key in provenance:
        if key not in PROVENANCE_KEYS:
            raise EnvelopeError("provenance-shape", f'unknown provenance field "{key}"')
    for key in REQUIRED_PROVENANCE_KEYS:
        if key not in provenance:
            raise EnvelopeError("provenance-shape", f'provenance is missing "{key}"')
    if provenance["origin"] not in ORIGINS:
        raise EnvelopeError(
            "provenance-shape",
            f'provenance.origin {json.dumps(provenance["origin"])} is not one of {", ".join(ORIGINS)}',
        )
    if not isinstance(provenance["synthetic"], bool):
        raise EnvelopeError("provenance-shape", "provenance.synthetic must be a boolean")
    stability = envelope.get("stability")
    if stability is not None:
        stable = _expect_object(stability, "stability-shape", "stability")
        for key in stable:
            if key not in STABILITY_KEYS:
                raise EnvelopeError("stability-shape", f'unknown stability field "{key}"')
        if "deterministic" in stable and not isinstance(stable["deterministic"], bool):
            raise EnvelopeError("stability-shape", "stability.deterministic must be a boolean")
        ignored = stable.get("ignored_fields")
        if ignored is not None and (not isinstance(ignored, list) or not all(isinstance(f, str) for f in ignored)):
            raise EnvelopeError("stability-shape", "stability.ignored_fields must be an array of strings")
    return envelope


def load_fixture_envelopes(fixtures_dir: Path, schemas_dir: Path) -> List[Dict[str, Any]]:
    registry = load_schema_registry(schemas_dir)
    seen: Dict[str, str] = {}
    loaded: List[Dict[str, Any]] = []
    for directory in sorted(Path(fixtures_dir).iterdir(), key=lambda p: p.name):
        fixture_file = directory / "fixture.json"
        if not directory.is_dir() or not fixture_file.is_file():
            continue
        try:
            raw = json.loads(fixture_file.read_text(encoding="utf-8"))
        except json.JSONDecodeError as error:
            raise EnvelopeError("json", f"{fixture_file} is not valid JSON: {error}") from error
        envelope = validate_envelope(raw, directory.name, registry)
        if envelope["fixture_id"] in seen:
            raise EnvelopeError(
                "duplicate-id",
                f'fixture_id "{envelope["fixture_id"]}" already defined by {seen[envelope["fixture_id"]]}',
            )
        seen[envelope["fixture_id"]] = str(fixture_file)
        loaded.append(envelope)
    return loaded


def summary_line(fixtures: List[Dict[str, Any]]) -> str:
    """Canonical summary string; must stay byte-identical across all loader implementations."""
    parts = [
        '{"fixture_id":"' + f["fixture_id"] + '","input_schema":"' + f["input"]["schema"] + '"}'
        for f in sorted(fixtures, key=lambda f: f["fixture_id"])
    ]
    return "[" + ",".join(parts) + "]"


def main() -> int:
    parser = argparse.ArgumentParser(description="Validate fixture envelopes; optionally print a FIXTURE_SUMMARY line.")
    parser.add_argument("--summary", action="store_true", help="print the canonical FIXTURE_SUMMARY line")
    parser.add_argument("fixtures_dir", nargs="?", default=None)
    parser.add_argument("schemas_dir", nargs="?", default=None)
    args = parser.parse_args()
    root = Path.cwd()
    fixtures_dir = Path(args.fixtures_dir) if args.fixtures_dir else root / "examples" / "fixtures"
    schemas_dir = Path(args.schemas_dir) if args.schemas_dir else root / "packages" / "schemas"
    try:
        fixtures = load_fixture_envelopes(fixtures_dir, schemas_dir)
    except EnvelopeError as error:
        print(error, file=sys.stderr)
        return 1
    if args.summary:
        print(f"FIXTURE_SUMMARY {summary_line(fixtures)}")
    else:
        for fixture in fixtures:
            print(f"{fixture['fixture_id']} ({fixture['kind']}) -> {fixture['input']['schema']}")
        print(f"{len(fixtures)} fixture(s) validated.")
    return 0


if __name__ == "__main__":
    raise SystemExit(main())

File: tests/fixtures-support/loaders/fixture-envelope.ts
/**
 * Cross-language fixture envelope loader (TypeScript reference implementation).
 *
 * Validation rules are mirrored in `fixture_envelope.py` and
 * `loaders/rust-loader/`. A rule change must land in all three at once.
 * Envelope contract: packages/schemas/fixture-envelope.v1.json (unified, ADR-017;
 * these loaders enforce the dataset-variant rules of that schema).
 * Payload validation is intentionally shallow (required + const at the top
 * level of the referenced canonical schema); deep validation belongs to the
 * schema toolchain, not to fixtures.
 *
 * Pure library: callers pass explicit fixture and schema directories.
 * Entry points keep process.argv handling out of this module.
 *
 * Erasable TypeScript only (runs on Node >= 22.6 type stripping).
 */
import { existsSync, readdirSync, readFileSync } from "node:fs";
import { join } from "node:path";

export const ENVELOPE_VERSION = "1.0";
const FIXTURE_ID_PATTERN = /^[a-z0-9]+(-[a-z0-9]+)*(\/[a-z0-9]+(-[a-z0-9]+)*)*$/;
const NAME_PATTERN = /^[a-z][a-z0-9-]*$/;
const TOP_LEVEL_KEYS = [
  "fixture_envelope_version",
  "fixture_id",
  "kind",
  "description",
  "input",
  "expected_outputs",
  "provenance",
  "stability",
];
const REQUIRED_TOP_LEVEL_KEYS = ["fixture_envelope_version", "fixture_id", "kind", "input", "provenance"];
const INPUT_KEYS = ["schema", "payload"];
const OUTPUT_KEYS = ["name", "schema", "payload"];
const PROVENANCE_KEYS = ["origin", "synthetic", "license", "notes"];
const REQUIRED_PROVENANCE_KEYS = ["origin", "synthetic"];
const ORIGINS = ["synthetic", "curated-public", "user-contributed"];
const STABILITY_KEYS = ["deterministic", "ignored_fields"];

export interface FixtureInput {
  schema: string;
  payload: Record<string, unknown>;
}
export interface ExpectedOutput {
  name: string;
  schema: string;
  payload: Record<string, unknown>;
}
export interface FixtureEnvelope {
  fixture_envelope_version: string;
  fixture_id: string;
  kind: string;
  description?: string;
  input: FixtureInput;
  expected_outputs?: ExpectedOutput[];
  provenance: Record<string, unknown>;
  stability?: Record<string, unknown>;
}
export interface LoadedFixture {
  envelope: FixtureEnvelope;
  inputSchema: unknown; // parsed canonical schema for the input payload
}
export interface SchemaRegistryEntry {
  filename: string;
  schema: Record<string, unknown>;
}
export type SchemaRegistry = Map<string, SchemaRegistryEntry>;

/** Fails with a stable `[rule]` prefix so tests and CI can attribute errors. */
function fail(rule: string, message: string): never {
  throw new Error(`[${rule}] ${message}`);
}

function expectObject(value: unknown, rule: string, what: string): Record<string, unknown> {
  if (typeof value !== "object" || value === null || Array.isArray(value)) {
    fail(rule, `${what} must be a JSON object`);
  }
  return value as Record<string, unknown>;
}

/** Strips the `.json` extension and trailing `.vN` from a schema $id path segment. */
export function baseSchemaName(schemaId: string): string {
  const segment = schemaId.split("/").filter(Boolean).pop() ?? "";
  return segment.replace(/\.json$/, "").replace(/\.v\d+$/, "");
}

export function loadSchemaRegistry(schemasDir: string): SchemaRegistry {
  const registry: SchemaRegistry = new Map();
  for (const name of readdirSync(schemasDir).sort()) {
    if (!name.endsWith(".json")) continue;
    const path = join(schemasDir, name);
    const parsed = JSON.parse(readFileSync(path, "utf8")) as Record<string, unknown>;
    const id = parsed.$id;
    if (typeof id === "string") {
      registry.set(id, { filename: name, schema: parsed });
    }
  }
  return registry;
}

function validatePayloadShallow(
  payload: unknown,
  schema: Record<string, unknown>,
  schemaId: string,
  what: string,
): void {
  const object = expectObject(payload, "payload-shape", what);
  const required = schema.required;
  if (Array.isArray(required)) {
    for (const key of required) {
      if (typeof key === "string" && !(key in object)) {
        fail("payload-required", `${what} is missing required field "${key}" of ${schemaId}`);
      }
    }
  }
  const properties = schema.properties;
  if (typeof properties === "object" && properties !== null) {
    for (const [key, definition] of Object.entries(properties as Record<string, unknown>)) {
      if (!(key in object) || typeof definition !== "object" || definition === null) continue;
      const constant = (definition as Record<string, unknown>).const;
      if (constant !== undefined && object[key] !== constant) {
        fail("payload-const", `${what} field "${key}" must equal ${JSON.stringify(constant)} per ${schemaId}`);
      }
    }
  }
}

function validateResolvablePayload(
  container: Record<string, unknown>,
  registry: SchemaRegistry,
  what: string,
): void {
  const schemaId = container.schema;
  if (typeof schemaId !== "string" || schemaId.length === 0) {
    fail("input-shape", `${what}.schema must be a non-empty string`);
  }
  const entry = registry.get(schemaId);
  if (!entry) {
    fail("schema-unresolved", `${what}.schema "${schemaId}" is not registered in packages/schemas`);
  }
  validatePayloadShallow(container.payload, entry.schema, schemaId, `${what}.payload`);
}

export function validateEnvelope(raw: unknown, fixtureDirName: string, registry: SchemaRegistry): FixtureEnvelope {
  const object = expectObject(raw, "shape", "fixture envelope");
  for (const key of Object.keys(object)) {
    if (!TOP_LEVEL_KEYS.includes(key)) {
      fail("unknown-field", `unknown top-level field "${key}"`);
    }
  }
  for (const key of REQUIRED_TOP_LEVEL_KEYS) {
    if (!(key in object)) {
      fail("missing-field", `missing required field "${key}"`);
    }
  }
  if (object.fixture_envelope_version !== ENVELOPE_VERSION) {
    fail("envelope-version", `fixture_envelope_version must be "${ENVELOPE_VERSION}", got ${JSON.stringify(object.fixture_envelope_version)}`);
  }
  const fixtureId = object.fixture_id;
  if (typeof fixtureId !== "string" || !FIXTURE_ID_PATTERN.test(fixtureId)) {
    fail("fixture-id", `fixture_id "${String(fixtureId)}" is not kebab-case`);
  }
  if (fixtureId !== fixtureDirName) {
    fail("fixture-id", `fixture_id "${fixtureId}" must equal its directory name "${fixtureDirName}"`);
  }
  const kind = object.kind;
  if (typeof kind !== "string" || !NAME_PATTERN.test(kind)) {
    fail("kind", `kind ${JSON.stringify(kind)} is not a lowercase kebab-case name`);
  }
  const input = expectObject(object.input, "input-shape", "input");
  for (const key of Object.keys(input)) {
    if (!INPUT_KEYS.includes(key)) fail("input-shape", `unknown input field "${key}"`);
  }
  for (const key of INPUT_KEYS) {
    if (!(key in input)) fail("input-shape", `input is missing "${key}"`);
  }
  if (typeof input.schema === "string" && baseSchemaName(input.schema) !== kind) {
    fail("kind-mismatch", `kind "${kind}" does not match input schema base name "${baseSchemaName(input.schema)}"`);
  }
  validateResolvablePayload(input, registry, "input");
  const outputs = object.expected_outputs;
  if (outputs !== undefined) {
    if (!Array.isArray(outputs)) fail("output-shape", "expected_outputs must be an array");
    for (const output of outputs) {
      const entry = expectObject(output, "output-shape", "expected_outputs entry");
      for (const key of Object.keys(entry)) {
        if (!OUTPUT_KEYS.includes(key)) fail("output-shape", `unknown expected_outputs field "${key}"`);
      }
      for (const key of OUTPUT_KEYS) {
        if (!(key in entry)) fail("output-shape", `expected_outputs entry is missing "${key}"`);
      }
      if (typeof entry.name !== "string" || !NAME_PATTERN.test(entry.name)) {
        fail("output-shape", `expected_outputs name ${JSON.stringify(entry.name)} is invalid`);
      }
      validateResolvablePayload(entry, registry, `expected_outputs "${String(entry.name)}"`);
    }
  }
  const provenance = expectObject(object.provenance, "provenance-shape", "provenance");
  for (const key of Object.keys(provenance)) {
    if (!PROVENANCE_KEYS.includes(key)) fail("provenance-shape", `unknown provenance field "${key}"`);
  }
  for (const key of REQUIRED_PROVENANCE_KEYS) {
    if (!(key in provenance)) fail("provenance-shape", `provenance is missing "${key}"`);
  }
  if (!ORIGINS.includes(String(provenance.origin))) {
    fail("provenance-shape", `provenance.origin ${JSON.stringify(provenance.origin)} is not one of ${ORIGINS.join(", ")}`);
  }
  if (typeof provenance.synthetic !== "boolean") {
    fail("provenance-shape", "provenance.synthetic must be a boolean");
  }
  const stability = object.stability;
  if (stability !== undefined) {
    const stable = expectObject(stability, "stability-shape", "stability");
    for (const key of Object.keys(stable)) {
      if (!STABILITY_KEYS.includes(key)) fail("stability-shape", `unknown stability field "${key}"`);
    }
    if (stable.deterministic !== undefined && typeof stable.deterministic !== "boolean") {
      fail("stability-shape", "stability.deterministic must be a boolean");
    }
    if (stable.ignored_fields !== undefined) {
      if (!Array.isArray(stable.ignored_fields) || stable.ignored_fields.some((f) => typeof f !== "string")) {
        fail("stability-shape", "stability.ignored_fields must be an array of strings");
      }
    }
  }
  return object as unknown as FixtureEnvelope;
}

export function loadFixtureEnvelopes(fixturesDir: string, schemasDir: string): LoadedFixture[] {
  const registry = loadSchemaRegistry(schemasDir);
  const seen = new Map<string, string>();
  const loaded: LoadedFixture[] = [];
  for (const name of readdirSync(fixturesDir).sort()) {
    const fixtureFile = join(fixturesDir, name, "fixture.json");
    if (!existsSync(fixtureFile)) continue;
    let raw: unknown;
    try {
      raw = JSON.parse(readFileSync(fixtureFile, "utf8"));
    } catch (error) {
      fail("json", `${fixtureFile} is not valid JSON: ${(error as Error).message}`);
    }
    const envelope = validateEnvelope(raw, name, registry);
    if (seen.has(envelope.fixture_id)) {
      fail("duplicate-id", `fixture_id "${envelope.fixture_id}" already defined by ${seen.get(envelope.fixture_id)}`);
    }
    seen.set(envelope.fixture_id, fixtureFile);
    loaded.push({
      envelope,
      inputSchema: registry.get(envelope.input.schema)?.schema,
    });
  }
  return loaded.sort((a, b) => (a.envelope.fixture_id < b.envelope.fixture_id ? -1 : 1));
}

/** Canonical summary string; must stay byte-identical across all loader implementations. */
export function summaryLine(fixtures: LoadedFixture[]): string {
  const parts = fixtures.map(
    (fixture) => `{"fixture_id":"${fixture.envelope.fixture_id}","input_schema":"${fixture.envelope.input.schema}"}`,
  );
  return `[${parts.join(",")}]`;
}
